File: lib/agents/win-loss.ts
import { searchWeb } from '../tools/serpapi';
import { scrapePage } from '../tools/firecrawl';
import { searchReddit, searchProductReviews } from '../tools/reddit';
import { searchHN } from '../tools/hn-algolia';
import { discoverAndScrape } from '../tools/discover-and-scrape';
import { planQueries } from '../tools/query-planner';
import { filterRelevant, requirementsFromContext } from '../tools/relevance';
import { extractCandidates, verifyCandidates } from '../tools/candidate-discovery';
import { generateHuggingFaceJson } from './gemini';
import type {
  AgentConfig,
  AgentContext,
  AgentOutput,
  WinLossOutput,
  WinReason,
  AgentSource,
  ConfidenceLevel,
  EvidenceCandidate,
} from './types';
import { scoreToLevel } from './types';
import { computeSignalQualityPenalty, extractToolResults } from '../tools/fallback';
import { isPlaceholderCompetitor, isUsableScrapePage, skippedScrapePromise } from './entity-url';
import { runResearchLoop } from './research-loop';
import { applyInsufficientGate, evidencePromptRules } from './evidence-gate';
import { localeFromGeography } from './search-locale';
import {
  resolveEntityProbes,
  resolveGapQueries,
  resolveSearchQueries,
} from './plan-queries';
import type { SearchResult, ScrapedPage, ToolResult } from '../tools/types';

function g2ReviewsUrl(competitorBrand: string): string {
  const slug = competitorBrand.toLowerCase().replace(/\s+/g, '-');
  return `https://www.g2.com/products/${slug}/reviews`;
}

async function run(ctx: AgentContext): Promise<AgentOutput> {
  const { query, product, competitor, priorContext, geography, category, namedEntities, requiredTerms } = ctx;

  const competitorName =
    competitor && !isPlaceholderCompetitor(competitor) ? competitor : 'discovered competitors';
  const g2Url =
    !isPlaceholderCompetitor(competitor) && competitor?.trim()
      ? g2ReviewsUrl(competitor)
      : null;
  // Skip G2 scrape when geography is set — G2 rarely covers local agritech
  const useG2 = Boolean(g2Url) && !geography;

  const locale = localeFromGeography(geography);
  const requirements = requirementsFromContext(ctx);

  const queryBundle = planQueries({
    product,
    competitor,
    domain: 'win-loss',
    query,
    category,
    geography,
    namedEntities,
    requiredTerms,
  });

  const planned = resolveSearchQueries(ctx, [queryBundle.broad, queryBundle.targeted], 2);
  const primaryQuery = planned[0] ?? queryBundle.broad;
  const secondaryQuery = planned[1] ?? queryBundle.hypothesis;
  const entityProbes = resolveEntityProbes(ctx, queryBundle.entityProbes, 2);
  const gapQueries = resolveGapQueries(ctx, 3);

  const searchedFor = [
    primaryQuery,
    secondaryQuery,
    queryBundle.hypothesis,
    ...entityProbes,
    ...gapQueries,
  ];

  let candidates: EvidenceCandidate[] = [];

  const loop = await runResearchLoop({
    domain: 'win-loss',
    requirements,
    budgetMs: 15_000,
    round1: () =>
      Promise.allSettled([
        searchWeb(primaryQuery, locale),
        searchProductReviews(product),
        isPlaceholderCompetitor(competitor)
          ? Promise.resolve({ data: [], timestamp: new Date().toISOString(), status: 'failed' as const, source: 'skip' })
          : searchProductReviews(competitorName),
        searchHN(`${product} ${geography?.name ?? ''} review`.trim()),
        useG2 && g2Url ? scrapePage(g2Url) : skippedScrapePromise(),
        searchWeb(secondaryQuery, locale),
        discoverAndScrape(queryBundle.hypothesis, {
          product,
          competitor: competitorName,
          domain: 'win-loss',
          topN: 2,
          keywords: queryBundle.keywords,
          locale,
          requirements,
        }),
        searchReddit(queryBundle.hypothesis),
        ...entityProbes.slice(0, 1).map(q => searchWeb(`${q} review feedback`, locale)),
      ]),
    ingest: (settled, round) => {
      const sources: AgentSource[] = [];
      const rawContent: string[] = [];
      const allHits: SearchResult[] = [];

      const takeSearch = (
        result: PromiseSettledResult<unknown> | undefined,
        label: string,
        tool: AgentSource['tool'],
        limit: number,
      ) => {
        if (!result || result.status !== 'fulfilled') return;
        const value = result.value as ToolResult<SearchResult[]>;
        if (!Array.isArray(value?.data)) return;
        // Reddit/HN posts may lack snippet — coerce
        const asSearch = value.data.map(r => ({
          title: r.title,
          url: r.url,
          snippet: ('snippet' in r ? String((r as SearchResult).snippet ?? '') : ''),
        }));
        const { kept } = filterRelevant(asSearch, requirements, {
          limit,
          minScore: geography ? 0.25 : 0.15,
        });
        kept.forEach(r => {
          allHits.push(r);
          const orig = value.data.find(d => d.url === r.url);
          sources.push({
            url: r.url,
            title: r.title,
            timestamp: (orig && 'created' in orig ? String((orig as { created: string }).created) : value.timestamp),
            tool,
          });
          rawContent.push(`[${label}] ${r.title}: ${r.snippet}`);
        });
      };

      if (round === 1) {
        takeSearch(settled[0], 'REVIEW SEARCH', 'serpapi', 5);
        takeSearch(settled[1], `REDDIT ${product}`, 'reddit', 4);
        takeSearch(settled[2], `REDDIT ${competitorName}`, 'reddit', 4);
        takeSearch(settled[3], 'HN', 'hn', 3);
        if (settled[4] && isUsableScrapePage(settled[4] as PromiseSettledResult<ToolResult<ScrapedPage>>) && competitor) {
          const page = (settled[4] as PromiseFulfilledResult<ToolResult<ScrapedPage>>).value.data;
          sources.push({
            url: page.url,
            title: `${competitor} — G2 reviews`,
            timestamp: (settled[4] as PromiseFulfilledResult<ToolResult<ScrapedPage>>).value.timestamp,
            tool: 'firecrawl',
          });
          rawContent.push(`[G2 REVIEWS] ${page.excerpt}`);
        }
        takeSearch(settled[5], 'SOCIAL REVIEW', 'serpapi', 3);
        if (settled[6]?.status === 'fulfilled') {
          const disc = settled[6].value as Awaited<ReturnType<typeof discoverAndScrape>>;
          disc.pages.forEach(pageResult => {
            if (pageResult.status === 'failed' || !pageResult.data.markdown?.trim()) return;
            sources.push({
              url: pageResult.data.url,
              title: pageResult.data.title || `${competitorName} review`,
              timestamp: pageResult.timestamp,
              tool: 'firecrawl',
            });
            rawContent.push(`[DISCOVERED REVIEW] ${pageResult.data.title}: ${pageResult.data.excerpt}`);
          });
          filterRelevant(disc.search.data, requirements, { limit: 5 }).kept.forEach(r => {
            allHits.push(r);
          });
        }
        takeSearch(settled[7], 'SALES REDDIT', 'reddit', 3);
        for (let i = 8; i < settled.length; i++) {
          takeSearch(settled[i], 'ENTITY REVIEW', 'serpapi', 3);
        }
        candidates = extractCandidates(allHits, {
          geographyName: geography?.name,
          exclude: [product, competitorName],
          limit: 5,
        });
      } else {
        for (const s of settled) {
          if (s.status !== 'fulfilled') continue;
          const val = s.value as Awaited<ReturnType<typeof discoverAndScrape>> | ToolResult<SearchResult[]>;
          if (val && 'pages' in val && Array.isArray(val.pages)) {
            val.pages.forEach(pageResult => {
              if (pageResult.status === 'failed' || !pageResult.data.markdown?.trim()) return;
              sources.push({
                url: pageResult.data.url,
                title: pageResult.data.title || 'Candidate',
                timestamp: pageResult.timestamp,
                tool: 'firecrawl',
              });
              rawContent.push(`[CANDIDATE REVIEW] ${pageResult.data.title}: ${pageResult.data.excerpt}`);
            });
          } else if (val && 'data' in val && Array.isArray(val.data)) {
            takeSearch(
              { status: 'fulfilled', value: val } as PromiseFulfilledResult<ToolResult<SearchResult[]>>,
              'GAP QUERY',
              'serpapi',
              4,
            );
          }
        }
      }

      return {
        sources,
        rawContent,
        toolResults: extractToolResults(settled as PromiseSettledResult<ToolResult<unknown>>[]),
        searchedFor,
        relevantSourceCount: sources.length,
        relevantHits: allHits,
        scrapedPageCount: sources.filter(s => s.tool === 'firecrawl').length,
      };
    },
    gapRound: async () => {
      if (gapQueries.length > 0) {
        const settled = await Promise.allSettled(
          gapQueries.slice(0, 2).map(q =>
            discoverAndScrape(q, {
              product,
              competitor: competitorName,
              domain: 'win-loss',
              topN: 1,
              keywords: queryBundle.keywords,
              locale,
              requirements,
            }),
          ),
        );
        searchedFor.push(...gapQueries.slice(0, 2));
        return settled;
      }

      const { settled, queries } = await verifyCandidates(candidates, {
        product,
        geographyName: geography?.name,
        category: category ?? 'reviews',
        maxCandidates: 2,
        topN: 1,
      });
      searchedFor.push(...queries);
      return settled;
    },
  });

  if (candidates.length) {
    loop.evidence = { ...loop.evidence, candidates };
  }

  const systemPrompt = `You are a win/loss analyst who reads buyer reviews to understand WHY deals are won or lost.
Focus on the BUYER perspective. Be specific — generic answers are useless.
${priorContext ? `\nPrior conversation context:\n${priorContext}` : ''}
${evidencePromptRules(geography, category)}`;

  const userPrompt = `Query: "${query}"
Our product: ${product}
Competitor: ${competitorName}
${geography ? `Geography: ${geography.name}` : ''}
Evidence status: ${loop.evidence.status}

Raw signals:
${(await import('@/lib/guardrails')).fenceUntrusted(loop.rawContent)}

Produce JSON:
{
  "insufficientEvidence": boolean,
  "facts": string[],
  "interpretation": string[],
  "competitorWins": [{ "reason": string, "frequency": "often" | "sometimes" | "rarely", "evidence": string }],
  "competitorLosses": [{ "reason": string, "frequency": "often" | "sometimes" | "rarely", "evidence": string }],
  "buyerSentiment": "positive" | "mixed" | "negative",
  "topSwitchTriggers": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}

If insufficientEvidence, leave wins/losses empty — do NOT cite unrelated G2 category pages.`;

  let parsed: Record<string, unknown> = {};
  try {
    parsed = await generateHuggingFaceJson(systemPrompt, userPrompt, {
      maxNewTokens: 1400,
      temperature: 0.2,
    });
  } catch {
    parsed = {
      insufficientEvidence: true,
      facts: [],
      interpretation: loop.evidence.gaps,
      competitorWins: [],
      competitorLosses: [],
      buyerSentiment: 'mixed',
      topSwitchTriggers: [],
      confidenceScore: 0.35,
    };
  }

  const gate = applyInsufficientGate(parsed, loop.evidence);
  if (gate.insufficient) loop.evidence = { ...loop.evidence, status: 'insufficient' };

  const confScore = Number.parseFloat(
    (gate.confidenceScore * computeSignalQualityPenalty(loop.toolResults, 7)).toFixed(2),
  );
  const confidence: ConfidenceLevel = scoreToLevel(confScore);

  const output: WinLossOutput = {
    agentId: 'win-loss',
    domain: 'win-loss',
    artifactType: 'win-loss-scorecard',
    confidence,
    confidenceScore: confScore,
    facts: gate.insufficient ? [] : (parsed.facts as string[] ?? []),
    interpretation: gate.insufficient ? loop.evidence.gaps : (parsed.interpretation as string[] ?? []),
    sources: loop.sources,
    generatedAt: new Date().toISOString(),
    competitor: competitorName,
    competitorWins: gate.insufficient ? [] : ((parsed.competitorWins ?? []) as WinReason[]),
    competitorLosses: gate.insufficient ? [] : ((parsed.competitorLosses ?? []) as WinReason[]),
    buyerSentiment: (parsed.buyerSentiment as WinLossOutput['buyerSentiment']) ?? 'mixed',
    topSwitchTriggers: gate.insufficient ? [] : (parsed.topSwitchTriggers as string[] ?? []),
    evidence: loop.evidence,
    toolCallCount: loop.toolCallCount,
    searchCallCount: loop.searchCallCount,
    scrapeCallCount: loop.scrapeCallCount,
    droppedIrrelevantCount: loop.droppedIrrelevantCount,
  };

  return output;
}

export const winLossAgent: AgentConfig = {
  id: 'win-loss',
  name: 'Win/Loss Agent',
  description: 'Reads G2 reviews, Reddit, and HN to surface why buyers choose one product over another.',
  run,
};

File: backend/app/agents/win_loss.py
"""Win / Loss Intelligence agent (parity with lib/agents/win-loss.ts)."""

from __future__ import annotations

import asyncio
import re
from datetime import datetime, timezone
from typing import Any

from app.agents.entity_url import is_placeholder_competitor, is_usable_scrape_page, skipped_scrape_promise
from app.gemini import generate_json
from app.models import score_to_level
from app.tools.fallback import compute_signal_quality_penalty, extract_tool_results
from app.tools.firecrawl import scrape_page
from app.tools.hn_algolia import search_hn
from app.tools.reddit import search_product_reviews, search_reddit
from app.tools.serpapi import search_web

AGENT_ID = "win-loss"
AGENT_NAME = "Win/Loss Agent"
AGENT_DESCRIPTION = (
    "Reads G2 reviews, Reddit, and HN to surface why buyers choose one product over another."
)


def _g2_reviews_url(competitor_brand: str) -> str:
    slug = competitor_brand.lower().replace(" ", "-")
    return f"https://www.g2.com/products/{slug}/reviews"


def _fulfilled(result: Any) -> bool:
    return not isinstance(result, BaseException)


async def run(ctx: dict[str, Any]) -> dict[str, Any]:
    query = ctx["query"]
    product = ctx["product"]
    competitor = ctx.get("competitor")
    prior_context = ctx.get("prior_context")

    competitor_name = competitor or "relevant competitors"
    g2_url = (
        _g2_reviews_url(competitor)
        if not is_placeholder_competitor(competitor) and (competitor or "").strip()
        else None
    )

    (
        web_result,
        reddit_product_result,
        reddit_competitor_result,
        hn_result,
        g2_scrape_result,
        social_review_result,
    ) = await asyncio.gather(
        search_web(f"{competitor_name} vs {product} review pros cons 2025"),
        search_product_reviews(product),
        search_product_reviews(competitor_name),
        search_hn(f"{competitor_name} {product} review comparison"),
        scrape_page(g2_url) if g2_url else skipped_scrape_promise(),
        search_web(
            f"{competitor_name} vs {product} "
            "site:x.com OR site:twitter.com OR site:instagram.com OR site:linkedin.com review comparison buyer feedback"
        ),
        return_exceptions=True,
    )

    (sales_reddit_result,) = await asyncio.gather(
        search_reddit(f"{product} review alternative experience"),
        return_exceptions=True,
    )

    sources: list[dict[str, Any]] = []
    raw_content: list[str] = []

    if _fulfilled(web_result):
        for r in (web_result.get("data") or [])[:5]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": web_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[REVIEW SEARCH] {r['title']}: {r['snippet']}")

    if _fulfilled(reddit_product_result):
        for p in (reddit_product_result.get("data") or [])[:4]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "reddit",
                }
            )
            raw_content.append(
                f"[REDDIT {product}] sentiment={p.get('sentiment')} | {p['title']}: {p['snippet']}"
            )

    if _fulfilled(reddit_competitor_result):
        for p in (reddit_competitor_result.get("data") or [])[:4]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "reddit",
                }
            )
            raw_content.append(
                f"[REDDIT {competitor_name}] sentiment={p.get('sentiment')} | {p['title']}: {p['snippet']}"
            )

    if _fulfilled(hn_result):
        for p in (hn_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "hn",
                }
            )
            raw_content.append(f"[HN] {p['title']}")

    if is_usable_scrape_page(g2_scrape_result) and competitor:
        page = g2_scrape_result["data"]
        sources.append(
            {
                "url": page["url"],
                "title": f"{competitor} — G2 reviews",
                "timestamp": g2_scrape_result["timestamp"],
                "tool": "firecrawl",
            }
        )
        raw_content.append(f"[G2 REVIEWS] {page['excerpt']}")

    if _fulfilled(social_review_result):
        for r in (social_review_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": r["url"],
                    "title": r["title"],
                    "timestamp": social_review_result["timestamp"],
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[SOCIAL REVIEW] {r['title']}: {r['snippet']}")

    if _fulfilled(sales_reddit_result):
        for p in (sales_reddit_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": p["url"],
                    "title": p["title"],
                    "timestamp": p["created"],
                    "tool": "reddit",
                }
            )
            raw_content.append(f"[SALES REDDIT] {p['title']}: {p['snippet']}")

    prior_block = f"\nPrior conversation context:\n{prior_context}" if prior_context else ""
    system_prompt = f"""You are a win/loss analyst who reads buyer reviews, Reddit discussions, and comparison content to understand WHY deals are won or lost. You look for patterns in real buyer language.

Rules:
- Focus on the BUYER perspective, not vendor claims.
- Separate facts (quoted from reviews) from interpretation.
- Be specific about reasons — generic answers are useless.
- Frequency: "often" = mentioned 3+ times, "sometimes" = 1-2 times, "rarely" = once.{prior_block}"""

    user_prompt = f"""Query: "{query}"
Our product: {product}
Competitor: {competitor_name}

Raw signals (buyer reviews, Reddit posts, comparisons):
{chr(10).join(raw_content)}

Produce JSON:
{{
  "facts": string[],
  "interpretation": string[],
  "competitorWins": [
    {{ "reason": string, "frequency": "often" | "sometimes" | "rarely", "evidence": string }}
  ],
  "competitorLosses": [
    {{ "reason": string, "frequency": "often" | "sometimes" | "rarely", "evidence": string }}
  ],
  "buyerSentiment": "positive" | "mixed" | "negative",
  "topSwitchTriggers": string[],
  "synthesizedAnswer": string,
  "confidenceScore": number
}}"""

    try:
        parsed = await generate_json(system_prompt, user_prompt, max_new_tokens=1400, temperature=0.2)
    except Exception:
        parsed = {
            "facts": [
                re.sub(r"^\[[^\]]+\]\s*", "", s)
                for s in raw_content[:3]
                if len(re.sub(r"^\[[^\]]+\]\s*", "", s)) > 15
            ],
            "interpretation": [
                "Analysis synthesis is temporarily unavailable. Raw data signals are shown below."
            ],
            "competitorWins": [],
            "competitorLosses": [],
            "buyerSentiment": "mixed",
            "topSwitchTriggers": [],
            "synthesizedAnswer": "Buyer sentiment data collected but synthesis failed.",
            "confidenceScore": 0.4,
        }

    raw_score = parsed["confidenceScore"] if isinstance(parsed.get("confidenceScore"), (int, float)) else 0.6
    tool_results = extract_tool_results(
        [
            web_result,
            reddit_product_result,
            reddit_competitor_result,
            hn_result,
            g2_scrape_result,
            social_review_result,
            sales_reddit_result,
        ]
    )
    conf_score = round(float(raw_score) * compute_signal_quality_penalty(tool_results, 7), 2)
    confidence = score_to_level(conf_score)

    return {
        "agentId": AGENT_ID,
        "domain": "win-loss",
        "artifactType": "win-loss-scorecard",
        "confidence": confidence,
        "confidenceScore": conf_score,
        "facts": parsed.get("facts") or [],
        "interpretation": parsed.get("interpretation") or [],
        "sources": sources,
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "competitor": competitor_name,
        "competitorWins": parsed.get("competitorWins") or [],
        "competitorLosses": parsed.get("competitorLosses") or [],
        "buyerSentiment": parsed.get("buyerSentiment") or "mixed",
        "topSwitchTriggers": parsed.get("topSwitchTriggers") or [],
    }
